osquery/tables/yara/yara.cpp

```diff
@@ -73,35 +114,82 @@ void doYARAScan(YR_RULES* rules,
   }
 }
 
-QueryData genYara(QueryContext& context) {
-  QueryData results;
+Status genYaraRuleFromFile(QueryContext& queryContext,
+                           YaraScanContext& scanContext) {
+  auto yaraParser = getYaraParser();
+  if (yaraParser == nullptr || yaraParser.get() == nullptr) {
+    return Status::failure("YARA config parser plugin has no pointer");
+  }
 
-  // Must specify a path constraint and at least one of sig_group or sigfile.
-  auto groups = context.constraints["sig_group"].getAll(EQUALS);
-  auto sigfiles = context.constraints["sigfile"].getAll(EQUALS);
-  if (groups.size() == 0 && sigfiles.size() == 0) {
-    return results;
+  auto& rules = yaraParser->rules();
+  auto sigfiles = queryContext.constraints["sigfile"].getAll(EQUALS);
+
+  for (const auto& file : sigfiles) {
+    // Check if this "ad-hoc" signature file has not been used/compiled.
+    if (rules.count(hashStr(file, YC_FILE)) == 0) {
```

Comment:
This is cached based on filename, right? Short of restarting osquery, is there a way to clear that cache? I don't think there needs to be, but the limitation might need documenting.

Reply:
The rule cache based on filename does not get cleanup. It avoids recompilation of file if they are already compiled.